File: 202. Happy Number/isHappy.java
public class Solution {
    public boolean isHappy(int n) {
        int sum = 0;
    	HashSet<Integer> cord = new HashSet<Integer>();
        String aaa = String.valueOf(Math.abs(n));
    	while (n != 1 && !cord.contains(sum)) {
    		cord.add(sum);
    		sum = 0;
    		int[] digit = new int[aaa.length()];
        	for (int i = 0; i < aaa.length(); i++) {
    			digit[i] = Integer.parseInt(String.valueOf(aaa.charAt(i)));
    		}
        	for (int i = 0; i < digit.length; i++) {
				sum += Math.pow(digit[i], 2);
			}
			n = sum;
// 			System.out.println("sum = "+sum);
        	aaa = String.valueOf(Math.abs(n));
		}
    	if (n == 1) {
			return true;
		}
    	else {
			return false;
		}
    }
}

File: 202. Happy Number/isHappy.py
class Solution(object):
    def isHappy_hash(self, n):
        """
        :type n: int
        :rtype: bool
        """
        cache = set()
        ops = lambda n : sum(int(x)**2 for x in str(n))
        while n not in cache:
            cache.add(n) 
            n = ops(n)
        return n == 1
            
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        ops = lambda n : sum(int(x)**2 for x in str(n))
        slow = n 
        fast = ops(n)
        while slow != fast:
            slow = ops(slow)
            fast = ops(ops(fast))
        return fast == 1